File: polyglot/c/injection_detector.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

// Define supported injection types
#define INJECTION_TYPES 5
typedef enum {
    SQL_INJECTION,
    XSS_INJECTION,
    COMMAND_INJECTION,
    LDAP_INJECTION,
    JSON_INJECTION
} InjectionType;

// Helper function to check for SQL injection patterns
int is_sql_injection(const char *input) {
    const char *patterns[] = {
        "SELECT", "INSERT", "UPDATE", "DELETE",
        "DROP", "TRUNCATE", "ALTER", "CREATE",
        "'", "\"", "--", ";", "UNION", "WHERE"
    };
    int i;
    for (i = 0; i < sizeof(patterns)/sizeof(patterns[0]); i++) {
        if (strstr(input, patterns[i])) {
            return 1;
        }
    }
    return 0;
}

// Helper function to check for XSS injection patterns
int is_xss_injection(const char *input) {
    const char *patterns[] = {
        "<script>", "</script>", "alert(", "onload=", "onclick=", "onerror="
    };
    int i;
    for (i = 0; i < sizeof(patterns)/sizeof(patterns[0]); i++) {
        if (strstr(input, patterns[i])) {
            return 1;
        }
    }
    return 0;
}

// Helper function to check for command injection patterns
int is_command_injection(const char *input) {
    const char *patterns[] = {
        ";", "|", "&", "(", ")", "{", "}", "[", "]", "$", "`", "||", "&&"
    };
    int i;
    for (i = 0; i < sizeof(patterns)/sizeof(patterns[0]); i++) {
        if (strstr(input, patterns[i])) {
            return 1;
        }
    }
    return 0;
}

// Helper function to check for LDAP injection patterns
int is_ldap_injection(const char *input) {
    const char *patterns[] = {
        "(", ")", "AND", "OR", "NOT", "=", "!="
    };
    int i;
    for (i = 0; i < sizeof(patterns)/sizeof(patterns[0]); i++) {
        if (strstr(input, patterns[i])) {
            return 1;
        }
    }
    return 0;
}

// Helper function to check for JSON injection patterns
int is_json_injection(const char *input) {
    const char *patterns[] = {
        "script", "<", ">", "&", "\"", "'", "{", "}"
    };
    int i;
    for (i = 0; i < sizeof(patterns)/sizeof(patterns[0]); i++) {
        if (strstr(input, patterns[i])) {
            return 1;
        }
    }
    return 0;
}

// Main injection detection function
InjectionType detect_injection(const char *input) {
    if (is_sql_injection(input)) {
        return SQL_INJECTION;
    } else if (is_xss_injection(input)) {
        return XSS_INJECTION;
    } else if (is_command_injection(input)) {
        return COMMAND_INJECTION;
    } else if (is_ldap_injection(input)) {
        return LDAP_INJECTION;
    } else if (is_json_injection(input)) {
        return JSON_INJECTION;
    }
    return 0;
}

// Helper function to convert InjectionType to string
const char* injection_type_to_string(InjectionType type) {
    switch (type) {
        case SQL_INJECTION: return "SQL Injection";
        case XSS_INJECTION: return "XSS Injection";
        case COMMAND_INJECTION: return "Command Injection";
        case LDAP_INJECTION: return "LDAP Injection";
        case JSON_INJECTION: return "JSON Injection";
        default: return "Unknown Injection";
    }
}

// Main entry point
int main() {
    char input[1024];

    printf("Enter input to audit for injection:\n");
    if (fgets(input, sizeof(input), stdin) != NULL) {
        input[strcspn(input, "\n")] = '\0'; // Remove newline

        InjectionType result = detect_injection(input);
        const char* result_str = injection_type_to_string(result);

        if (result != 0) {
            printf("Injection detected: %s\n", result_str);
        } else {
            printf("No injection detected.\n");
        }
    } else {
        printf("Error reading input.\n");
    }

    return 0;
}

File: polyglot/java/injection_detector.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.*;

public class InjectionDetector {
    private static final Set<String> SQL_KEYWORDS = new HashSet<>(Arrays.asList(
        "SELECT", "INSERT", "UPDATE", "DELETE", "DROP", "ALTER", "CREATE", "TRUNCATE",
        "UNION", "OR", "AND", "NOT", "IN", "LIKE", "BETWEEN", "CASE", "WHEN", "THEN",
        "ELSE", "ISNULL", "IF", "WHERE", "FROM", "JOIN", "ON", "HAVING", "GROUP", "ORDER"
    ));

    private static final Set<String> XSS_KEYWORDS = new HashSet<>(Arrays.asList(
        "<script>", "</script>", "<img", "<a", "<form", "<input", "<body", "<html",
        "<style>", "<meta", "<link", "<iframe", "<frame", "<table", "<tr", "<td"
    ));

    private static final Set<String> CMD_KEYWORDS = new HashSet<>(Arrays.asList(
        ";", "|", "&", "&&", "||", ">", "<", "<<", ">>", "(", ")", "{", "}", "[", "]",
        "!", "$", "`", "exec", "system", "run", "call"
    ));

    private static final Set<String> HEADER_KEYWORDS = new HashSet<>(Arrays.asList(
        "Content-Type", "Authorization", "Cookie", "User-Agent", "Referer", "Host"
    ));

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: java InjectionDetector <file_path>");
            return;
        }

        String filePath = args[0];
        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
            String line;
            int lineNumber = 0;
            while ((line = reader.readLine()) != null) {
                lineNumber++;
                detectInjections(line, lineNumber);
            }
        } catch (IOException e) {
            System.err.println("Error reading file: " + e.getMessage());
        }
    }

    private static void detectInjections(String line, int lineNumber) {
        boolean sqlDetected = false;
        boolean xssDetected = false;
        boolean cmdDetected = false;
        boolean headerDetected = false;

        // SQL Injection Detection
        for (String keyword : SQL_KEYWORDS) {
            if (line.contains(keyword)) {
                sqlDetected = true;
                break;
            }
        }

        // XSS Injection Detection
        for (String keyword : XSS_KEYWORDS) {
            if (line.contains(keyword)) {
                xssDetected = true;
                break;
            }
        }

        // Command Injection Detection
        for (String keyword : CMD_KEYWORDS) {
            if (line.contains(keyword)) {
                cmdDetected = true;
                break;
            }
        }

        // Header Injection Detection
        for (String keyword : HEADER_KEYWORDS) {
            if (line.contains(keyword)) {
                headerDetected = true;
                break;
            }
        }

        if (sqlDetected || xssDetected || cmdDetected || headerDetected) {
            System.out.println("Line " + lineNumber + ": Possible injection detected");
            if (sqlDetected) System.out.println("  - SQL Injection keywords found");
            if (xssDetected) System.out.println("  - XSS Injection keywords found");
            if (cmdDetected) System.out.println("  - Command Injection keywords found");
            if (headerDetected) System.out.println("  - Header Injection keywords found");
        }
    }
}